File: tools/core070_namespace_1_batch.R
# Retained evidence runner for the "namespace-1" manifest-area batch, Tier 0:
# existence / registration parity (48 EXECUTABLE_NOW + 6 NEEDS_NEW_JULIA_SURFACE
# cases; see docs/dev-log/core070/namespace-1-batch-contract.json for the full
# triage, including the 36 REUSED_OR_RECLASSIFY rows this batch deliberately
# does not fabricate checks for).
#
# This tier needs NO installed gllvmTMB package and NO frozen library: every
# check is a text scan of the PINNED readback R source tree (source_pins in
# the contract), confirming that each row's cited S3 method / export is both
# (a) registered in the readback NAMESPACE and (b) actually defined at the
# cited file. That is why this runner takes only a destination directory --
# unlike tools/core070_postfit_policy_batch.R, there is no <frozen-library>
# argv slot here (see the contract's runner.tier1_followup for the deferred
# numeric-fit tier that WOULD need one).
#
# argv (matches the contract's runner.outer_argv and this file's own
# --self-test, which never reaches disk under that path):
#   Rscript --vanilla tools/core070_namespace_1_batch.R <destination>
#
# <destination> must not already exist; it is created and holds r-facts.json,
# julia-facts.json, julia-stdout.log, julia-stderr.log, results.tsv,
# diagnostics.log, and receipt.json.

args <- commandArgs(TRUE)
self_test <- length(args) == 1L && identical(args[[1]], "--self-test")

sha256_file <- function(path) {
  command <- if (nzchar(Sys.which("sha256sum"))) "sha256sum" else "shasum"
  argv <- if (identical(command, "sha256sum")) path else c("-a", "256", path)
  line <- system2(command, argv, stdout = TRUE, stderr = TRUE)
  stopifnot(is.null(attr(line, "status")), length(line) >= 1L)
  sub("[[:space:]].*$", "", line[[1L]])
}

suppressPackageStartupMessages(library(jsonlite))

root <- normalizePath(".")
contract_path <- file.path(root, "docs/dev-log/core070/namespace-1-batch-contract.json")
if (!file.exists(contract_path)) {
  stop("FATAL: contract not found at ", contract_path, " -- refusing to run with no state.")
}
contract <- jsonlite::read_json(contract_path, simplifyVector = FALSE)
contract_sha256 <- sha256_file(contract_path)

stopifnot(
  identical(contract$reference_commit, "b4d5fee64def88bc768dda1f1f77c29b295edd86"),
  identical(contract$status, "FROZEN_NAMESPACE_1_BATCH_CONTRACT"),
  length(contract$cases) == contract$expected_case_count,
  contract$expected_case_count == 48L,
  length(contract$needs_new_julia_surface) == contract$needs_new_julia_surface_count,
  contract$needs_new_julia_surface_count == 6L,
  length(contract$negative_controls) >= 2L
)

# --- validate pinned R source (NAMESPACE + every cited file) ---------------
source_root <- file.path(root, ".unlazy/core070-aghq/oracle-source/readback")
namespace_path <- file.path(root, contract$namespace_pin)
stopifnot(file.exists(namespace_path))
namespace_sha_actual <- sha256_file(namespace_path)
stopifnot(identical(namespace_sha_actual, contract$namespace_sha256))
namespace_lines <- readLines(namespace_path, warn = FALSE)

for (rel in names(contract$source_pins)) {
  path <- file.path(source_root, rel)
  stopifnot(file.exists(path))
  digest <- sha256_file(path)
  stopifnot(identical(digest, contract$source_pins[[rel]]))
}

# ---------------------------------------------------------------------------
# The one check this tier performs: for one row, is the NAMESPACE line
# present verbatim, and does the cited file contain a line matching the
# row's definition regex? Exposed as a function so --self-test can call it
# directly on deliberately mutated inputs (rejected-mutation checks).
# ---------------------------------------------------------------------------
check_row <- function(row, ns_lines, file_lines) {
  registered <- row$r_namespace_line %in% trimws(ns_lines)
  defined <- any(grepl(row$r_definition_pattern, file_lines, perl = TRUE))
  list(registered = registered, defined = defined)
}

# Cache file contents once per cited file.
file_cache <- new.env(parent = emptyenv())
get_file_lines <- function(rel) {
  if (!exists(rel, envir = file_cache, inherits = FALSE)) {
    assign(rel, readLines(file.path(source_root, rel), warn = FALSE), envir = file_cache)
  }
  get(rel, envir = file_cache, inherits = FALSE)
}

all_rows <- c(contract$cases, contract$needs_new_julia_surface)

run_all_checks <- function(ns_lines, cache_fn) {
  out <- list()
  for (row in all_rows) {
    fl <- cache_fn(row$r_file)
    out[[row$case_id]] <- check_row(row, ns_lines, fl)
  }
  out
}

r_results <- run_all_checks(namespace_lines, get_file_lines)

r_facts <- lapply(all_rows, function(row) {
  res <- r_results[[row$case_id]]
  list(
    case_id = row$case_id,
    source_id = row$source_id,
    registered = res$registered,
    defined = res$defined,
    ok = isTRUE(res$registered) && isTRUE(res$defined)
  )
})
names(r_facts) <- vapply(all_rows, `[[`, "", "case_id")

# ---------------------------------------------------------------------------
# --self-test: exercise check_row() against the real pinned sources (no
# destination dir, no Julia child) AND against >=4 deliberately mutated
# inputs to prove each check actually discriminates rather than vacuously
# passing. FAILS LOUDLY (non-zero exit via stop()) on any mismatch.
# ---------------------------------------------------------------------------
if (self_test) {
  mismatches <- character(0)
  for (row in all_rows) {
    res <- r_results[[row$case_id]]
    if (!isTRUE(res$registered) || !isTRUE(res$defined)) {
      mismatches <- c(mismatches, sprintf(
        "case %s: expected registered=TRUE defined=TRUE, got registered=%s defined=%s",
        row$case_id, res$registered, res$defined))
    }
  }

  # --- rejected mutations (>=4): corrupt inputs the check must catch -------
  rejected <- character(0)
  probe_row <- all_rows[[1]]
  probe_lines <- get_file_lines(probe_row$r_file)

  # 1. NAMESPACE line missing entirely.
  ns_without <- namespace_lines[trimws(namespace_lines) != trimws(probe_row$r_namespace_line)]
  m1 <- check_row(probe_row, ns_without, probe_lines)
  if (isTRUE(m1$registered)) stop("REJECTED MUTATION 1 FAILED TO BE CAUGHT: NAMESPACE-line removal still read as registered")
  rejected <- c(rejected, "NAMESPACE line removed -> registered=FALSE (caught)")

  # 2. Definition line removed from the source file.
  lines_without_def <- probe_lines[!grepl(probe_row$r_definition_pattern, probe_lines, perl = TRUE)]
  m2 <- check_row(probe_row, namespace_lines, lines_without_def)
  if (isTRUE(m2$defined)) stop("REJECTED MUTATION 2 FAILED TO BE CAUGHT: definition removal still read as defined")
  rejected <- c(rejected, "definition line removed -> defined=FALSE (caught)")

  # 3. NAMESPACE line present but with a typo'd class name (must NOT match).
  garbled_ns_line <- sub(",", ",ZZZ_", probe_row$r_namespace_line, fixed = FALSE)
  ns_garbled <- c(namespace_lines, garbled_ns_line)
  ns_garbled <- ns_garbled[trimws(ns_garbled) != trimws(probe_row$r_namespace_line)]
  m3 <- check_row(probe_row, ns_garbled, probe_lines)
  if (isTRUE(m3$registered)) stop("REJECTED MUTATION 3 FAILED TO BE CAUGHT: garbled NAMESPACE line still read as registered")
  rejected <- c(rejected, "garbled NAMESPACE class name -> registered=FALSE (caught)")

  # 4. Definition present under a DIFFERENT (wrong) generic name (must NOT match).
  wrong_def <- gsub(probe_row$r_name, paste0("zzz_", probe_row$r_name), probe_lines, fixed = TRUE)
  m4 <- check_row(probe_row, namespace_lines, wrong_def)
  if (isTRUE(m4$defined)) stop("REJECTED MUTATION 4 FAILED TO BE CAUGHT: renamed-generic definition still read as defined")
  rejected <- c(rejected, "definition renamed to a different generic -> defined=FALSE (caught)")

  if (length(rejected) < 4L) stop("FATAL: fewer than 4 rejected mutations ran")

  # --- negative controls (>=2): rows whose Julia symbol is genuinely absent,
  #     confirmed via the paired Julia self-test's own contract read (this R
  #     process does not load Julia; it asserts the CONTRACT records these as
  #     expected_julia_symbol_exists == FALSE, which is what the Julia
  #     self-test independently verifies against the live GLLVM module).
  neg_case_ids <- vapply(contract$negative_controls, function(x) {
    if (is.null(x$case_id)) NA_character_ else x$case_id
  }, character(1))
  neg_case_ids <- neg_case_ids[!is.na(neg_case_ids)]
  if (length(neg_case_ids) < 2L) stop("FATAL: fewer than 2 case-anchored negative controls in contract")
  for (cid in neg_case_ids) {
    row <- Filter(function(r) identical(r$case_id, cid), contract$needs_new_julia_surface)[[1]]
    if (!identical(row$expected_julia_symbol_exists, FALSE)) {
      stop("NEGATIVE CONTROL ", cid, " FAILED TO BE CAUGHT: contract does not mark it as an expected absence")
    }
  }

  if (length(mismatches)) {
    for (m in mismatches) message("SELF_TEST_MISMATCH: ", m)
    stop("FATAL: R self-test found ", length(mismatches), " mismatch(es) on the real pinned sources -- see above.")
  }

  cat("CORE070_NAMESPACE_1_R_SELF_TEST_OK rows=", length(all_rows),
      " rejected_mutations=", length(rejected),
      " negative_controls=", length(neg_case_ids), "\n", sep = "")
  quit(status = 0L)
}

# ---------------------------------------------------------------------------
# Normal (non-self-test) invocation: write r-facts.json, invoke the Julia
# child, cross-check, write receipt.
# ---------------------------------------------------------------------------
stopifnot(length(args) == 1L)
output_dir <- args[[1]]
stopifnot(!dir.exists(output_dir))
dir.create(output_dir, recursive = TRUE)

r_facts_path <- file.path(output_dir, "r-facts.json")
jsonlite::write_json(
  list(schema = "core070-namespace-1-r-facts/v1", status = "OK", facts = r_facts),
  r_facts_path, auto_unbox = TRUE, pretty = TRUE, null = "null"
)

julia_out <- file.path(output_dir, "julia-facts.json")
julia_bin <- Sys.which("julia")
stopifnot(nzchar(julia_bin))
t0 <- Sys.time()
julia_status <- system2(julia_bin, c("--project=.", "tools/core070_namespace_1_batch.jl", julia_out),
                         stdout = file.path(output_dir, "julia-stdout.log"),
                         stderr = file.path(output_dir, "julia-stderr.log"))
elapsed <- as.numeric(Sys.time() - t0, units = "secs")

julia_report <- if (file.exists(julia_out)) {
  jsonlite::read_json(julia_out, simplifyVector = FALSE)
} else NULL

exec_ids <- vapply(contract$cases, `[[`, "", "case_id")
needs_ids <- vapply(contract$needs_new_julia_surface, `[[`, "", "case_id")

julia_ok <- function(row) {
  if (is.null(julia_report)) return(FALSE)
  f <- julia_report$facts[[row$julia_symbol]]
  !is.null(f) && identical(f$exists, row$expected_julia_symbol_exists)
}

exec_pass <- vapply(contract$cases, function(row) {
  r_ok <- isTRUE(r_facts[[row$case_id]]$ok)
  r_ok && julia_ok(row)
}, logical(1))
names(exec_pass) <- exec_ids

needs_pass <- vapply(contract$needs_new_julia_surface, function(row) {
  r_ok <- isTRUE(r_facts[[row$case_id]]$ok)  # R side is still expected to be registered+defined
  r_ok && julia_ok(row)                       # and the Julia absence/partial fact must match expectation
}, logical(1))
names(needs_pass) <- needs_ids

results_ok <- !is.null(julia_report) &&
  identical(julia_status, 0L) &&
  identical(julia_report$status, "OK") &&
  all(exec_pass) &&
  all(needs_pass)

raw_lines <- c(
  vapply(exec_ids, function(id) paste(id, if (isTRUE(exec_pass[[id]])) "PASS" else "FAIL", "executable_now", sep = "\t"), character(1)),
  vapply(needs_ids, function(id) paste(id, if (isTRUE(needs_pass[[id]])) "PASS" else "FAIL", "needs_new_julia_surface", sep = "\t"), character(1))
)
raw_path <- file.path(output_dir, "results.tsv")
writeLines(raw_lines, raw_path)

diag_lines <- character(0)
if (!identical(julia_status, 0L)) diag_lines <- c(diag_lines, paste("julia_exit_code", julia_status))
if (is.null(julia_report)) diag_lines <- c(diag_lines, "julia-facts.json was not written")
diag_path <- file.path(output_dir, "diagnostics.log")
writeLines(diag_lines, diag_path)

receipt <- list(
  status = if (isTRUE(results_ok)) "PASS" else "FAIL",
  scope = "CORE070_NAMESPACE_1_BATCH_TIER0",
  reference_commit = contract$reference_commit,
  contract_sha256 = contract_sha256,
  namespace_sha256 = namespace_sha_actual,
  source_pins = contract$source_pins,
  source_unchanged = TRUE,
  executable_now_count = length(exec_ids),
  needs_new_julia_surface_count = length(needs_ids),
  reused_or_reclassify_count = contract$reused_or_reclassify_count,
  julia_exit_code = julia_status,
  julia_elapsed_seconds = elapsed,
  julia_facts_sha256 = if (file.exists(julia_out)) sha256_file(julia_out) else NA_character_,
  r_facts_sha256 = sha256_file(r_facts_path),
  raw_sha256 = sha256_file(raw_path),
  diagnostics_sha256 = sha256_file(diag_path),
  r_version = R.version.string
)
receipt_path <- file.path(output_dir, "receipt.json")
jsonlite::write_json(receipt, receipt_path, auto_unbox = TRUE, pretty = TRUE, null = "null")

cat("CORE070_NAMESPACE_1_BATCH_", receipt$status, "\n", sep = "")
quit(status = if (isTRUE(results_ok)) 0L else 1L)

File: tools/core070_namespace_1_batch.jl
# Julia child for the "namespace-1" manifest-area batch (Tier 0: existence /
# registration parity -- see docs/dev-log/core070/namespace-1-batch-contract.json
# for the full 48-EXECUTABLE_NOW / 6-NEEDS_NEW_JULIA_SURFACE / 36-REUSED_OR_RECLASSIFY
# triage and the deferred Tier 1 numeric-parity follow-up).
#
# This process carries NO RCall dependency and reads NO R state: it loads
# GLLVM.jl itself and answers, for every symbol named in the contract's
# `cases` + `needs_new_julia_surface` lists, "does this symbol exist at
# GLLVM module scope?" via `isdefined`. That is the entirety of what this
# tier checks -- the outer R runner (tools/core070_namespace_1_batch.R)
# answers the matching R-side question by scanning the pinned readback R
# source text, with no installed gllvmTMB package required either.
#
# argv (matches the contract's runner.inner_argv and this file's own
# --self-test, which never reaches disk under that path):
#   julia --project=. tools/core070_namespace_1_batch.jl <destination>/julia-facts.json
#
# <destination> must already exist (the outer R runner mkpath()s it before
# invoking this child). This script also mkpath()s the output's parent
# directory defensively in case it is ever invoked standalone.
#
# No external JSON dependency (JSON3 is not a declared project dependency);
# reuses the hand-rolled minimal JSON reader/writer already established as
# repo convention in tools/core070_postfit_policy_batch.jl / core070_postfit_1_batch.jl.

using GLLVM

# ---------------------------------------------------------------------------
# Minimal JSON reader/writer (no external dependency; mirrors the existing
# repo convention in tools/core070_postfit_policy_batch.jl).
# ---------------------------------------------------------------------------
function json_read(path::AbstractString)
    txt = read(path, String)
    pos = Ref(1)
    skip_ws!() = begin
        while pos[] <= lastindex(txt) && isspace(txt[pos[]])
            pos[] = nextind(txt, pos[])
        end
    end
    local parse_value
    function parse_string()
        pos[] += 1
        buf = IOBuffer()
        while true
            c = txt[pos[]]
            if c == '"'
                pos[] = nextind(txt, pos[])
                break
            elseif c == '\\'
                pos[] = nextind(txt, pos[])
                e = txt[pos[]]
                if e == 'n'; write(buf, '\n')
                elseif e == 't'; write(buf, '\t')
                elseif e == 'u'
                    hex = txt[nextind(txt, pos[]):nextind(txt, pos[], 4)]
                    write(buf, Char(parse(UInt32, hex; base = 16)))
                    pos[] = nextind(txt, pos[], 4)
                else write(buf, e)
                end
                pos[] = nextind(txt, pos[])
            else
                write(buf, c)
                pos[] = nextind(txt, pos[])
            end
        end
        String(take!(buf))
    end
    function parse_number()
        start = pos[]
        while pos[] <= lastindex(txt) && (isdigit(txt[pos[]]) || txt[pos[]] in ('-', '+', '.', 'e', 'E'))
            pos[] = nextind(txt, pos[])
        end
        s = txt[start:prevind(txt, pos[])]
        return occursin('.', s) || occursin('e', s) || occursin('E', s) ? parse(Float64, s) : parse(Int, s)
    end
    function parse_array()
        pos[] += 1
        out = Any[]
        skip_ws!()
        if txt[pos[]] == ']'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            push!(out, parse_value())
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == ']'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    function parse_object()
        pos[] += 1
        out = Dict{String, Any}()
        skip_ws!()
        if txt[pos[]] == '}'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            k = parse_string()
            skip_ws!()
            @assert txt[pos[]] == ':'
            pos[] = nextind(txt, pos[])
            skip_ws!()
            out[k] = parse_value()
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == '}'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    parse_value = function ()
        skip_ws!()
        c = txt[pos[]]
        if c == '"'; return parse_string()
        elseif c == '['; return parse_array()
        elseif c == '{'; return parse_object()
        elseif c == 't'; pos[] += 4; return true
        elseif c == 'f'; pos[] += 5; return false
        elseif c == 'n'; pos[] += 4; return nothing
        else; return parse_number()
        end
    end
    skip_ws!()
    parse_value()
end

json_escape(s::AbstractString) = replace(s, "\\" => "\\\\", "\"" => "\\\"", "\n" => "\\n")
to_json(x::Bool) = x ? "true" : "false"
to_json(x::Nothing) = "null"
to_json(x::AbstractString) = "\"" * json_escape(x) * "\""
to_json(x::Integer) = string(x)
to_json(x::AbstractFloat) = isfinite(x) ? repr(x) : "null"
to_json(x::AbstractVector) = "[" * join(to_json.(x), ",") * "]"
to_json(x::AbstractDict) = "{" * join(("\"$(json_escape(string(k)))\":" * to_json(v) for (k, v) in x), ",") * "}"

const CONTRACT_PATH = joinpath(@__DIR__, "..", "docs/dev-log/core070/namespace-1-batch-contract.json")

function load_contract()
    isfile(CONTRACT_PATH) ||
        error("FATAL: contract not found at $CONTRACT_PATH -- refusing to run with no state.")
    return json_read(CONTRACT_PATH)
end

# The one check this tier performs, exposed as a function so --self-test can
# call it directly without touching disk.
symbol_exists(mod::Module, sym::AbstractString) = isdefined(mod, Symbol(sym))

const SYNTHETIC_NEG_SYMBOL = "gllvmTMB_julia_bridge_nonexistent_surface_zzz"

function collect_facts(contract)
    symbols = Set{String}()
    for c in contract["cases"]
        push!(symbols, c["julia_symbol"])
    end
    for c in contract["needs_new_julia_surface"]
        push!(symbols, c["julia_symbol"])
    end
    push!(symbols, SYNTHETIC_NEG_SYMBOL)

    facts = Dict{String, Any}()
    for sym in symbols
        facts[sym] = Dict{String, Any}("exists" => symbol_exists(GLLVM, sym))
    end
    return facts
end

function main()
    length(ARGS) == 1 ||
        error("FATAL: expected exactly 1 argv (output path), got $(length(ARGS)): $(ARGS)")
    out_path = ARGS[1]
    isfile(out_path) && error("FATAL: destination already exists: $out_path")
    out_dir = dirname(out_path)
    isempty(out_dir) || mkpath(out_dir)

    contract = load_contract()
    facts = collect_facts(contract)

    report = Dict{String, Any}(
        "schema" => "core070-namespace-1-julia-facts/v1",
        "status" => "OK",
        "julia_version" => string(VERSION),
        "symbol_count" => length(facts),
        "facts" => facts,
    )

    open(out_path, "w") do io
        write(io, to_json(report))
    end
    println("CORE070_NAMESPACE_1_JULIA_FACTS_OK symbols=", length(facts))
end

# --self-test exercises collect_facts() against the real contract and the
# real GLLVM module (no file I/O to a destination), and FAILS LOUDLY (throws,
# nonzero exit) rather than silently reporting a partial/empty result if
# GLLVM does not load or the contract is missing -- matching the hardened
# template's "verifier FAILS LOUDLY on missing state even under --self-test"
# rule, applied here to the runner side of that same discipline.
function self_test()
    contract = load_contract()
    facts = collect_facts(contract)

    n_expect_true = count(c -> c["expected_julia_symbol_exists"], contract["cases"])
    n_expect_false = count(c -> !c["expected_julia_symbol_exists"], contract["needs_new_julia_surface"])

    mismatches = String[]
    for c in contract["cases"]
        sym = c["julia_symbol"]
        expect = c["expected_julia_symbol_exists"]
        got = get(get(facts, sym, Dict()), "exists", nothing)
        if got !== expect
            push!(mismatches, "case $(c["case_id"]): expected exists=$expect for $sym, got $got")
        end
    end
    for c in contract["needs_new_julia_surface"]
        sym = c["julia_symbol"]
        expect = c["expected_julia_symbol_exists"]
        got = get(get(facts, sym, Dict()), "exists", nothing)
        if got !== expect
            push!(mismatches, "needs-row $(c["case_id"]): expected exists=$expect for $sym, got $got")
        end
    end
    got_neg = get(get(facts, SYNTHETIC_NEG_SYMBOL, Dict()), "exists", nothing)
    if got_neg !== false
        push!(mismatches, "synthetic true-negative symbol resolved exists=$got_neg, expected false")
    end

    if !isempty(mismatches)
        for m in mismatches
            println(stderr, "SELF_TEST_MISMATCH: ", m)
        end
        error("FATAL: self-test found $(length(mismatches)) mismatch(es) -- see stderr. " *
              "n_expect_true=$n_expect_true n_expect_false=$n_expect_false symbols_collected=$(length(facts))")
    end

    println("CORE070_NAMESPACE_1_JULIA_SELF_TEST_OK exec_true=", n_expect_true,
            " needs_false=", n_expect_false, " total_symbols=", length(facts))
end

if abspath(PROGRAM_FILE) == @__FILE__
    if length(ARGS) == 1 && ARGS[1] == "--self-test"
        self_test()
    else
        main()
    end
end
